### ISSUE TITLE
traefik example using etcd

### ISSUE THREAD

**Issue by minrk** [jupyterhub/outreachy#20]
Extending #7, it would be good to have an example configuration of traefik using etcd to store the routing table.

The goal: achieve the same result as #7, but use etcd to store configuration instead of placing it directly in the traefik.toml configuration.

Tasks:

1. install and start etcd
2. create (manually) the appropriate etcd values
3. pass the same tests in #19 using the new configuration

References:

- [etcd](https://coreos.com/etcd/)
- [traefik etcd](https://docs.traefik.io/configuration/backends/etcd/)

**Comment by GeorgianaElena** [comment]
Hi! I would love to work on this issue. :)

**Comment by minrk** [comment]
I thought you might :) Please go ahead!

**Comment by GeorgianaElena** [comment]
Haha, I'm so unpredictable.

**Comment by GeorgianaElena** [comment]
Hi @minrk! I ran into some problems when trying to run the tests.

I installed and started etcd as a local standalone cluster. I managed to add to this etcd cluster both manually (through `etcdctl put` commands) but also using traefik's storeconfig capability, the configuration in the toml file.

When running traefik from the cmd `traefik --etcd --etcd.endpoint=127.0.0.1:2379 --etcd.useapiv3=true --etcd.watch=true`, the routing happens as it should. However, when I try to launch the proc from the pytest using `Popen`, I'm receiving this warning from etcd:
```
WARNING: 2018/10/11 16:38:55 grpc: Server.processUnaryRPC failed to write status: connection error: desc = "transport is closing"
2018-10-11 16:44:15.892360 D | etcdserver/api/v3rpc: failed to receive lease keepalive request from gRPC stream ("rpc error: code = Canceled desc = context canceled")
```
And the request to traefik fails with a 404.

Do you have any idea why this is happening?

**Comment by GeorgianaElena** [comment]
Indeed it was happening too quickly. There were two things happening there which caused the weird behavior:
1. When I switched traefik to use etcd it started generating more logs than before. Because I was redirecting the traefik stdout to Suprocess.PIPE, sometimes the pipe would get full causing my processes to deadlock. This one is solved now.

2. Now that etcd is involved, traefik becomes ready slower than before. My checkHostUp assert passes (i.e. traefik accepts new connections, thus it should be ready), but the subsequent requests that the test sends fail sometimes with 404 because traefik hasn't had the change yet to fully load the config from etcd. I couldn't find any guarantees in the documentation about the order of these events, so now I'm trying to figure out when traefik is actually ready (accepts connections + etcd config is loaded).

Thanks a lot for your help. I will get back with updates.

**Comment by minrk** [comment]
Implemented in #27
